### ISSUE TITLE
Symbols hierarchy

### ISSUE THREAD

**Issue by laughedelic** [scalameta/language-server#33]
Currently symbol type is used in place of its hierarchical parent reference:

https://github.com/scalameta/language-server/blob/04ff21002da48ea293dbed9ad99efbb366198d0e/metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala#L197-L200

Citing [LSP spec](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#document-symbols-request) (`textDocument/documentSymbol` response):

```
	/**
	 * The name of the symbol containing this symbol. This information is for
	 * user interface purposes (e.g. to render a qaulifier in the user interface
	 * if necessary). It can't be used to re-infer a hierarchy for the document
	 * symbols.
	 */
	containerName?: string;
```

So symbols get grouped by type instead of the declaration hierarchy. I don't know if there is such thing in the VS Code UI, but in Atom there is an Outline view panel which directly corresponds to the `textDocument/documentSymbol` feature:

![screen shot 2017-11-12 at 17 37 43](https://user-images.githubusercontent.com/766656/32701129-3ead1f96-c7d0-11e7-8f29-32051f5b4eb3.png)

And here is an example Flow JS source:

![screen shot 2017-11-12 at 17 34 55](https://user-images.githubusercontent.com/766656/32701134-4a52fdb6-c7d0-11e7-80e4-337a1801ccb4.png)

I'd like to fix it, but I don't know how to get this hierarchical information from the SemanticDB.

**Comment by olafurpg** [comment]
Wow, that would be super cool to support! scala.meta.Symbol is an ADT with cases

- Symbol.None
- Symbol.Local, these are definitions inside local blocks, can't be globally referenced
- Symbol.Global, this one is the one you're looking for, you can create hierarchical data structures from it with `.owner/.signature`
- Symbol.Multi

Source https://github.com/scalameta/scalameta/blob/23de1ca839e53ad6387981bd7cced35122a2cf76/langmeta/shared/src/main/scala/org/langmeta/semanticdb/Symbol.scala#L24

To get the SymbolKind of a symbol you need the `Denotation` from https://github.com/scalameta/language-server/blob/04ff21002da48ea293dbed9ad99efbb366198d0e/metaserver/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala#L23

**Comment by laughedelic** [comment]
OK. Thanks for the references. I'll submit a PR shortly. Didn't solve it completely, but got something.

**Comment by olafurpg** [comment]
You probably have to change the signature of SymbolIndexer.documentSymbols to include the Symbol.Global from here

https://github.com/scalameta/language-server/blob/04ff21002da48ea293dbed9ad99efbb366198d0e/metaserver/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala#L36

and then use `Symbol.Global.owner.syntax` as the container name.

**Comment by laughedelic** [comment]
Thanks. I actually moved the implementation to `SymbolIndexer.documentSymbols`  and changed its return type to `Seq[SymbolInformation]`. Is it ok or should I be more conservative with the interface?

**Comment by laughedelic** [comment]
Symbols's `syntax` is definitely not what we want here because `name` is what is shown directly in the GUI, and `containerName` has to refer to the same `name` parent symbol has.

**Comment by olafurpg** [comment]
Fixed by #34 @laughedelic ?

**Comment by laughedelic** [comment]
Yes, let's close it. The nested hierarchy issue is resolved by #34. It still needs improvements, but it will be addressed in another PR.
